File: tools/crosscheck/crosscheck.R
# Cross-check: recompute the Python engine's deterministic outputs in halosimr
# and assert an exact match. Reads expected.json (from crosscheck.py), rebuilds
# the same fixed inputs, and compares the exposure matrix, per-provider gap
# statistics, and the four readiness kernels (no warmup).
#
# Run:  Rscript crosscheck.R expected.json
# Exits non-zero on any mismatch.

args <- commandArgs(trailingOnly = TRUE)
json_path <- if (length(args)) args[1] else "expected.json"

suppressPackageStartupMessages({
  library(halosimr)
  library(jsonlite)
})

exp <- jsonlite::fromJSON(json_path, simplifyMatrix = TRUE)
N_DAYS    <- exp$n_days
PROVIDERS <- exp$providers
THRESHOLD <- exp$threshold

# --- rebuild the identical inputs (mirror of build_inputs in crosscheck.py) ---
sched <- matrix("o", nrow = 4, ncol = N_DAYS)
for (t in 0:(N_DAYS - 1)) {
  j <- t + 1L
  if (t %% 3 == 0) sched[1, j] <- "d"
  if (t %% 3 == 1) sched[2, j] <- "n"
  if (t %% 2 == 0) sched[3, j] <- "d"
}
events <- data.frame(
  day_idx = c(0L, 6L, 12L, 18L, 1L, 7L, 13L),
  shift_type = c(rep("day", 4), rep("night", 3)),
  stringsAsFactors = FALSE
)

sim <- list(schedule = sched, events = events, providers = PROVIDERS,
            readiness_threshold_days = THRESHOLD,
            readiness_half_life_days = exp$half_life, ebbinghaus_b = exp$ebb_b,
            step_t2_days = exp$step_t2, step_partial_value = exp$step_partial)

# ---------------------------------------------------------------------------
fails <- 0L
report <- function(label, ok) {
  cat(sprintf("%-40s %s\n", label, if (isTRUE(ok)) "MATCH" else "MISMATCH"))
  if (!isTRUE(ok)) fails <<- fails + 1L
}

# --- exposure matrix ---
res <- halosimr:::compute_exposure(sim)
exposure_r <- matrix(as.integer(res$exposure_matrix), nrow = nrow(res$exposure_matrix))
report("exposure matrix", identical(exposure_r, matrix(as.integer(exp$exposure),
                                                       nrow = nrow(exp$exposure))))

# --- per-provider gap statistics ---
rdf <- res$results_df
er  <- exp$results
num_cols <- c("n_events", "gap_min", "gap_q25", "gap_median", "gap_mean",
              "gap_q75", "gap_max")
gaps_ok <- all(vapply(num_cols, function(cn)
  isTRUE(all.equal(as.numeric(rdf[[cn]]), as.numeric(er[[cn]]), tolerance = 1e-9)),
  logical(1)))
report("gap statistics (numeric)", gaps_ok)

# t2first: Python null -> NA; compare with NAs aligned
t2_ok <- identical(is.na(rdf$t2first), is.na(er$t2first)) &&
  isTRUE(all.equal(rdf$t2first[!is.na(rdf$t2first)],
                   as.integer(er$t2first[!is.na(er$t2first)])))
report("gap time-to-first (t2first)", t2_ok)

flag_ok <- identical(as.logical(rdf$max_gap_exceeds_threshold),
                     as.logical(er$max_gap_exceeds_threshold))
report("gap exceeds-threshold flag", flag_ok)

# --- readiness kernels (no warmup, combined = exposure) ---
sim$combined_matrix <- res$exposure_matrix
for (m in c("binary", "exponential", "ebbinghaus", "step")) {
  sim$readiness_model <- m
  r_r <- halosimr:::compute_readiness(sim, initial_last = NULL)
  r_py <- matrix(as.numeric(exp$readiness[[m]]), nrow = nrow(exp$readiness[[m]]))
  report(sprintf("readiness: %s", m),
         isTRUE(all.equal(unname(r_r), unname(r_py), tolerance = 1e-9)))
}

cat("\n")
if (fails == 0L) {
  cat("ALL STAGES MATCH: halosimr reproduces the halosim engine exactly.\n")
} else {
  cat(sprintf("%d STAGE(S) MISMATCHED.\n", fails))
  quit(status = 1)
}

File: tools/crosscheck/crosscheck.py
"""Cross-check: dump the halosim (Python) engine's deterministic outputs.

Feeds a fixed, hand-built schedule and event stream (no RNG) through the Python
engine and writes the exposure matrix, per-provider gap statistics, and the four
readiness kernels (computed with no warmup) to expected.json. crosscheck.R then
recomputes the same quantities in halosimr and asserts an exact match.

RNG-dependent stages (schedule/event generation, warmup) are deliberately
excluded: the two engines match by design on the deterministic computation, not
on random draws.

Run:  python3 crosscheck.py > expected.json
      (with the halosim package importable, e.g. PYTHONPATH=projects/halosim)
"""
import json
import numpy as np
import pandas as pd

from halosim.simulation import Simulation
from halosim.exposure import compute_exposure
from halosim.readiness import compute_readiness

N_DAYS = 20
PROVIDERS = ["P1", "P2", "P3", "P4"]
THRESHOLD = 8          # readiness threshold / step t1 / gap flag
HALF_LIFE = 6
EBB_B = 0.05
STEP_T2 = 14
STEP_PARTIAL = 0.5
MODELS = ["binary", "exponential", "ebbinghaus", "step"]


def build_inputs():
    """Deterministic schedule + events, constructed identically in R."""
    sched = np.full((4, N_DAYS), "o", dtype="U1")
    for t in range(N_DAYS):
        if t % 3 == 0:
            sched[0, t] = "d"      # P1: day shift every 3rd day
        if t % 3 == 1:
            sched[1, t] = "n"      # P2: night shift on t % 3 == 1
        if t % 2 == 0:
            sched[2, t] = "d"      # P3: day shift on even days
        # P4: always off
    events = pd.DataFrame(
        [{"day_idx": d, "shift_type": "day"} for d in (0, 6, 12, 18)]
        + [{"day_idx": d, "shift_type": "night"} for d in (1, 7, 13)]
    )
    return sched, events


def native(x):
    """Convert numpy scalars / NaN to JSON-native values."""
    if isinstance(x, (np.integer,)):
        return int(x)
    if isinstance(x, (np.floating,)):
        return None if np.isnan(x) else float(x)
    if isinstance(x, (np.bool_,)):
        return bool(x)
    if x is None or (isinstance(x, float) and np.isnan(x)):
        return None
    return x


def main():
    sched, events = build_inputs()
    sim = Simulation(
        n_days=N_DAYS, providers=list(PROVIDERS), schedule=sched, events=events,
        readiness_model="binary", readiness_threshold_days=THRESHOLD,
        readiness_half_life_days=HALF_LIFE, ebbinghaus_b=EBB_B,
        step_partial_value=STEP_PARTIAL, step_t2_days=STEP_T2,
    )

    exposure_matrix, results_df = compute_exposure(sim)

    # Readiness with no warmup, combined = exposure (no training), per model.
    sim.combined_matrix = exposure_matrix
    readiness = {}
    for m in MODELS:
        sim.readiness_model = m
        r = compute_readiness(sim, initial_last=None)
        readiness[m] = r.tolist()

    results = [
        {k: native(v) for k, v in rec.items()}
        for rec in results_df.to_dict(orient="records")
    ]

    out = {
        "n_days": N_DAYS,
        "providers": PROVIDERS,
        "threshold": THRESHOLD,
        "half_life": HALF_LIFE,
        "ebb_b": EBB_B,
        "step_t2": STEP_T2,
        "step_partial": STEP_PARTIAL,
        "exposure": exposure_matrix.astype(int).tolist(),
        "results": results,
        "readiness": readiness,
    }
    print(json.dumps(out, indent=2))


if __name__ == "__main__":
    main()
